File: Leetcode/src/_1_TwoSum.cc
#include <unordered_map>
#include <vector>

class Solution {
 public:
  std::vector<int> twoSum(std::vector<int> &nums, int target) {
    std::size_t n = nums.size();
    std::unordered_map<int, size_t> numToIndex{};
    numToIndex.reserve(n);

    for (std::size_t i = 0UZ; i < n; i++) {
      auto it = numToIndex.find(target - nums[i]);
      if (it != numToIndex.end()) [[unlikely]] {
        return {static_cast<int>(i), static_cast<int>(it->second)};
      }
      numToIndex.insert({nums[i], i});
    }

    return {};
  }
};

File: Leetcode/src/bin/_1_TwoSum.java
import java.util.HashMap;

class Solution {
  public int[] twoSum(int[] nums, int target) {
    int n = nums.length;

    HashMap<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < n; i++) {
      int x = nums[i];
      if (map.containsKey(target - x)) {
        return new int[] { map.get(target - x), i };
      }
      map.put(x, i);
    }

    return new int[] {};
  }
}
